### PR TITLE
Drop root privileges in docker container

### PR DESCRIPTION
**Pull request by DerOetzi**
Implements changes for #259 

Changes:

- Run service as user habapp which is member of group habapp
- new environment parameters USER_ID and GROUP_ID for the user (defaults to 9001 like openhab docker image)
- Shrink docker image size by using explicit build image
- Add some more python standard  files and directories to .gitignore (using GitHub default .gitignore template for python projects)

Looking forward to your feedback

### CODE REVIEW

**Comment by spacemanspiff2007** [comment]
su-exec is available for all [architectures](https://pkgs.alpinelinux.org/packages?name=su-exec&branch=edge).
What was the reason to use gosu instead of su-exec?

**Comment by DerOetzi** [comment]
Just old Habits 😉 But you'r right will make things easier to use. Will Change this

File: Dockerfile [hunk 1]
```diff
@@ -1,25 +1,46 @@
...
+		tini \
+# Support for Timezones
+    	tzdata; \
+    mkdir -p ${HABAPP_HOME}; \
+    mkdir -p ${HABAPP_HOME}/config; \
+# install HABApp
+    pip3 install \
+        --no-index \
+        --find-links=/root/wheels \
+		habapp; \
```
Comment (spacemanspiff2007) [inline Dockerfile:36]: Can you please install from the sources and not from pip?

File: Dockerfile [hunk 2]
```diff
@@ -1,25 +1,46 @@
...
+FROM python:3.9-alpine as buildimage
 
-VOLUME [ "/config"]
+COPY . /tmp/app_install
 
-# Install required dependencies
-RUN apk add --no-cache \
-# Support for Timezones
-    tzdata \
+RUN set -eux;\
```
[the remark this answers is not in the document]
Comment (DerOetzi) [reply Dockerfile:5]: -e Ensures the script is even aborted on simple errors
-u Writes an error message to stderr if a used env variable is empty
-x every command is expanded and printed to stdout before execution. This make it clearer to debug when you reduce docker layers by chaining commands in one RUN command

More details you can find here: https://explainshell.com/explain?cmd=set+-eux

File: Dockerfile [hunk 3]
```diff
@@ -1,25 +1,46 @@
...
+
+RUN set -eux; \
+# Install required dependencies
+ 	apk add --no-cache \
+    	bash \
+		su-exec \
+		tini \
+# Support for Timezones
+    	tzdata; \
+    mkdir -p ${HABAPP_HOME}; \
```
[the remark this answers is not in the document]
Comment (DerOetzi) [reply Dockerfile:30]: Volumes are treated like every linux mount. For stability reasons the directory should exist in filesystem before mounting anything to it.

File: Dockerfile (hunk 1 above)
Comment (DerOetzi) [reply Dockerfile:36]: I think you mean not to install from pypi?!

This is not the case here the complete local src is copied to the buildimage (line 3) and a wheel is build and put together with all dependencies wheels from pypi (line 11) Those are copied to /root/wheels of the final container (line 15) the  combination of --no-index  and --find-links tells pip to install from this local files not from pypi

Here is the log output of the command

```bash
+ pip3 install --no-index '--find-links=/root/wheels' habapp
Looking in links: /root/wheels
Processing /root/wheels/HABApp-0.31.3.dev0-py3-none-any.whl
Processing /root/wheels/aiohttp_sse_client-0.2.1-py2.py3-none-any.whl
Processing /root/wheels/paho_mqtt-1.6.1-py3-none-any.whl
Processing /root/wheels/ujson-5.0.0-cp39-cp39-musllinux_1_1_x86_64.whl
Processing /root/wheels/voluptuous-0.12.2-py3-none-any.whl
Processing /root/wheels/aiohttp-3.8.1-cp39-cp39-musllinux_1_1_x86_64.whl
Processing /root/wheels/stackprinter-0.2.5-py3-none-any.whl
Processing /root/wheels/EAScheduler-0.1.4-py3-none-any.whl
Processing /root/wheels/EasyCo-0.2.3-py3-none-any.whl
Processing /root/wheels/immutables-0.16-cp39-cp39-linux_x86_64.whl
Processing /root/wheels/pendulum-2.1.2-cp39-cp39-linux_x86_64.whl
Processing /root/wheels/pydantic-1.8.2-py3-none-any.whl
Processing /root/wheels/watchdog-2.1.6-py3-none-any.whl
Processing /root/wheels/bidict-0.21.4-py3-none-any.whl
Processing /root/wheels/frozenlist-1.2.0-cp39-cp39-musllinux_1_1_x86_64.whl
Processing /root/wheels/attrs-21.4.0-py2.py3-none-any.whl
Processing /root/wheels/multidict-5.2.0-cp39-cp39-musllinux_1_1_x86_64.whl
Processing /root/wheels/async_timeout-4.0.2-py3-none-any.whl
Processing /root/wheels/aiosignal-1.2.0-py3-none-any.whl
Processing /root/wheels/charset_normalizer-2.0.10-py3-none-any.whl
Processing /root/wheels/yarl-1.7.2-cp39-cp39-musllinux_1_1_x86_64.whl
Processing /root/wheels/astral-2.2-py2.py3-none-any.whl
Processing /root/wheels/pytz-2021.3-py2.py3-none-any.whl
Processing /root/wheels/ruamel.yaml-0.17.20-py3-none-any.whl
Processing /root/wheels/pytzdata-2020.1-py2.py3-none-any.whl
Processing /root/wheels/python_dateutil-2.8.2-py2.py3-none-any.whl
Processing /root/wheels/typing_extensions-4.0.1-py3-none-any.whl
Processing /root/wheels/six-1.16.0-py2.py3-none-any.whl
Processing /root/wheels/idna-3.3-py3-none-any.whl
Processing /root/wheels/ruamel.yaml.clib-0.2.6-cp39-cp39-linux_x86_64.whl
```

File: Dockerfile [hunk 4]
```diff
@@ -1,25 +1,46 @@
...
+		g++; \
+# wheel all packages for habapp
+	cd /tmp/app_install; \
+	pip wheel --wheel-dir=/root/wheels --use-feature=in-tree-build .
 
-# Always use latest versions
-RUN mkdir -p /usr/src/app
-WORKDIR /usr/src/app
-COPY . .
+FROM python:3.9-alpine
```
Comment (spacemanspiff2007) [inline Dockerfile:13]: Why is this used again and not buildimage?

File: Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (spacemanspiff2007) [reply Dockerfile:30]: Got it - thanks!

**Comment by spacemanspiff2007** [comment]
I am not that firm with docker so I am trusting you with the changes. 😉 
If you could reformat the file a bit (leading spaces) then it's imho good to be merged.
Thank you for you explanations.

File: Dockerfile (hunk 4 above)
Comment (DerOetzi) [reply Dockerfile:13]: This is the beginning of the final container. Everything added inside the buildimage part wan't be available to the final container unless you copy it with COPY --from=buildimage. For example in the final container image there will be no /tmp/app_install directory and no g++ and it's dependencies is installed. This way the size of the final container image is reduced a lot.

**Comment by DerOetzi** [comment]
> I am not that firm with docker so I am trusting you with the changes. wink If you could reformat the file a bit (leading spaces) then it's imho good to be merged. Thank you for you explanations.
> 
> One more thing: What would be a good mechanism to install custom packages?



> If you could reformat the file a bit (leading spaces) then it's imho good to be merged. 

Sorry didn't recognize that my code formatter for Dockerfiles isn't working. Hope I fixed all.

> Thank you for you explanations.

Your very welcome thanks for this create project.

> One more thing: What would be a good mechanism to install custom packages?

What exactly do you mean custom packages? Packages an user of HABApp might need inside rules beside habapp?
In this case I would say the most common way with docker would be to use your image as base-image and install whatever is needed on top.

```Dockerfile
FROM spacemanspiff2007/habapp:latest as buildimage

RUN pip3 wheel ... # pip wheel everything you might need

FROM spacemanspiff2007/habapp:latest

COPY --from=buildimage /root/wheels /root/wheels

RUN pip3 install ... # pip install all packages from wheels
```
There might be one problem that you use the -alpine image as base image for your container, because some very common python packages like scipy or numpy or pandas have no ready wheels on pipy for Alpine. That's means those have to be build from scratch inside the container, which will take several hours each time you build the container. For this reason maybe it would be a good idea to switch to the -slim python base image which is a little bit bigger then the -alpine one but has supported wheels for nearly every most common python packages because it is a debian based base image.

**Comment by spacemanspiff2007** [comment]
If you think it's a good idea we can switch to the slim image.

**Comment by DerOetzi** [comment]
> If you think it's a good idea we can switch to the slim image.

Will do the changes this evening.

**Comment by DerOetzi** [comment]
I did the switch to slim baseimage. The resulting image is about 70MB bigger then the alpine one which is in my opinion okay, for giving more compatibilities for different common python packages

**Comment by spacemanspiff2007** [comment]
Great - thank you very much!
Could you additionally add a small step by step guide under [Docker]https://github.com/spacemanspiff2007/HABApp/blob/master/_doc/installation.rst#installation-1) how to use the image as base image and add own packages from pip.

**Comment by DerOetzi** [comment]
Hopefully I will find the time at the weekend

**Comment by DerOetzi** [comment]
I have seen that you rst files for documentation. I never worked with this and don't have time to work into it. I can right you an .md file und you may transform it if this is okay for you?

**Comment by spacemanspiff2007** [comment]
Yes - of course! Let's do it that way!

**Comment by DerOetzi** [comment]
I have added docker.md to the container directory. You may although use this for docker hub description. If you still have questions please contact me. I can offer that I review your changes you will make out of my input.

**Comment by spacemanspiff2007** [comment]
Thank you very much for this huge contribution!

**Comment by spacemanspiff2007** [comment]
I've added a space before ``\`` and replaced the hard coded path with the env variable.
Additionally I'm initializing the GROUP_ID from USER_ID to match the docs.

@DerOetzi 
Could you check if this still runs fine?
Would it be possible to start HABApp out of the ``/config`` dir instead of ``HABAPP_HOME``?
Then the image change would be backwards compatible.

**Pull request merged**